atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/expectations/KeyValueServiceDataTracker.java

```diff
@@ -0,0 +1,129 @@
+    private final LongAdder kvsCallsOverall = new LongAdder();
+    private final ConcurrentMap<TableReference, KvsCallReadInfo> maximumBytesKvsCallInfoByTable =
+            new ConcurrentHashMap<>();
+    private final AtomicReference<Optional<KvsCallReadInfo>> maximumBytesKvsCallInfoOverall =
+            new AtomicReference<>(Optional.empty());
+
+    public TransactionReadInfo getReadInfo() {
+        return ImmutableTransactionReadInfo.builder()
+                .bytesRead(bytesReadOverall.longValue())
+                .kvsCalls(kvsCallsOverall.longValue())
+                .maximumBytesKvsCallInfo(maximumBytesKvsCallInfoOverall.get())
+                .build();
+    }
+
+    /**
+     * This is un-synchronized as it is called after task completion and the abort/commit stage.
+     * For returned statistics, no consistency guarantees are provided if the user interacts with the transaction
+     * outside the context of {@link com.palantir.atlasdb.transaction.api.TransactionTask} or after commit/abort
+     * stages (e.g. this can happen with futures/iterators).
+     */
+    public ImmutableMap<TableReference, TransactionReadInfo> getReadInfoByTable() {
+        Set<TableReference> tableRefs = ImmutableSet.copyOf(Sets.intersection(
+                bytesReadByTable.asMap().keySet(), kvsCallByTable.asMap().keySet()));
+
+        return tableRefs.stream()
+                .collect(ImmutableMap.toImmutableMap(
+                        Function.identity(), tableRef -> ImmutableTransactionReadInfo.builder()
+                                .bytesRead(bytesReadByTable.get(tableRef))
+                                .kvsCalls(kvsCallByTable.get(tableRef))
+                                .maximumBytesKvsCallInfo(
+                                        Optional.ofNullable(maximumBytesKvsCallInfoByTable.get(tableRef)))
+                                .build()));
+    }
+
+    /**
+     * Tracks an effectively completed kvs read method call for some {@link TableReference}.
+     * Effectively completed refers to an eager call (i.e. does not spawn futures or iterators for later consumption).
+     */
+    public void readForTable(TableReference tableRef, String methodName, long bytesRead) {
+        KvsCallReadInfo callInfo = ImmutableKvsCallReadInfo.builder()
+                .bytesRead(bytesRead)
+                .methodName(methodName)
+                .build();
+        updateKvsMethodOverallTallies(callInfo);
+        updateKvsMethodByTableTallies(tableRef, callInfo);
+    }
+
+    /**
+     * Track a lazy kvs read method call for some {@link TableReference}.
+     */
+    public void partialReadForTable(TableReference tableRef, long bytesRead) {
+        bytesReadOverall.add(bytesRead);
+        bytesReadByTable.addAndGet(tableRef, bytesRead);
+    }
+
+    /**
+     * Track that a kvs read method was called for some {@link TableReference}.
+     */
+    public void callForTable(TableReference tableRef) {
```

Comment:
The expected pattern is that clients must call `callForTable` before calling partialRead.

Consider: callForTable returns some functional interface that consumers can use to record additional bytes read. That method can also use the method name context, and in turn can do its own aggregation to determine the total amount of bytes read (although it maybe misleading if all those bytes read weren't in a single call to the KVS...)

Reply:
Separately, if each partial read makes a new call to the KVS, do you want those internal calls to be tracked?

Reply:
What do you mean by the last question?